File: java_old/gameBomberman/src/main/java/ru/atom/game/http/matchmaker/MatchMakerController.java
package ru.atom.game.http.matchmaker;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import ru.atom.game.databases.player.PlayerDataRepository;
import ru.atom.game.socket.util.JsonHelper;

import java.util.Map;

@Controller
@RequestMapping("matchmaker")
public class MatchMakerController {
    private static final Logger log = LoggerFactory.getLogger(MatchMakerController.class);

    @Autowired
    private MatchMaker matchMaker;

    @Autowired
    private PlayerDataRepository playerDataRepository;

    @RequestMapping(
            path = "casual",
            method = RequestMethod.POST,
            produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<String> join(@RequestBody String userData) {
        ResponseEntity<String> response;
        Map<String, Object> data = JsonHelper.fromJson(userData, Map.class);
        String name = ((Map) data.get("credentials")).get("name").toString();

        if((response = auth(name)) != null) {
            return response;
        }
        Integer id = matchMaker.getCommonSessionID(name);
        return new ResponseEntity<>(id.toString(), HttpStatus.OK);
    }

    @RequestMapping(
            path = "rating",
            method = RequestMethod.POST,
            produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<String> rating(@RequestBody String userData) {
        ResponseEntity<String> response;
        Map<String, Object> data = JsonHelper.fromJson(userData, Map.class);
        String name = ((Map) data.get("credentials")).get("name").toString();

        if((response = auth(name)) != null) {
            return response;
        }

        Integer id = matchMaker.getRatingSessionID(name);
        return new ResponseEntity<>(id.toString(), HttpStatus.OK);
    }

    private ResponseEntity<String> auth(String name) {
        /*
        List<PlayerData> player = playerDataRepository.findByName(name);
        if(player.size() == 0) {
            return new ResponseEntity<>("Cant find this user", HttpStatus.NOT_FOUND);
        }
        */
        return null;
    }
}

File: MatchMaker/src/main/kotlin/io/rybalkinsd/kotlinbootcamp/server/MatchMakerController.kt
package io.rybalkinsd.kotlinbootcamp.server.matchmaker

import io.rybalkinsd.kotlinbootcamp.server.Game
import io.rybalkinsd.kotlinbootcamp.server.HttpClient
import io.rybalkinsd.kotlinbootcamp.util.logger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RequestParam
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

@Controller
@RequestMapping("/chat")
class MatchMakerController {
    val log = logger()

    val games: Queue<Game> = ConcurrentLinkedQueue()

    @Autowired
    lateinit var client: HttpClient

    // todo read about spring security
    // todo read about spring services
    // todo read about services communication
    // todo read about oath2
    // todo read about encryption
    // todo implement basis
    @RequestMapping(
            path = ["/play"],
            method = [RequestMethod.POST],
            consumes = [MediaType.APPLICATION_FORM_URLENCODED_VALUE]
    )
    fun play(@RequestParam("name") name: String): ResponseEntity<String> {
        val checkResult = checkUser(name)
        if (checkResult != null)
            return checkResult
        var game = games.poll()
        if (game == null) {
            val gameId = getGameId(2)
                    ?: return ResponseEntity.status(500).body("Failed to get Game instance Id")
            game = Game(gameId)
        }
        game.currentPlayerAmount++
        if (game.isReady())
            startGame(game.id)
        else
            games.add(game)

        return ResponseEntity.ok(game.id.toString())
    }

    private fun startGame(id: Int) {
        client.startGame(id)
    }

    private fun getGameId(count: Int): Int? {
        return client.getGameId(count)
    }

    fun checkUser(name: String): ResponseEntity<String>? = when {
        name.length < 3 -> ResponseEntity.badRequest().body("Too Short Name")
        name.length > 20 -> ResponseEntity.badRequest().body("Too Long Name")
        else -> null
    }
}
